src/ServiceControl/MessageRedirects/Api/MessageRedirectsModule.cs

```diff
@@ -0,0 +1,169 @@
+                });
+
+                return HttpStatusCode.Created;
+            };
+
+            Put["/redirects/{messageredirectid:guid}"] = parameters =>
+            {
+                Guid messageRedirectId = parameters.messageredirectid;
+
+                var request = BindAndNormalize();
+
+                var result = ValidateRedirectRequest(request, false);
+
+                if (result.NegotiationContext.StatusCode != HttpStatusCode.OK)
+                {
+                    return result;
+                }
+
+                var message = new ChangeMessageRedirect
+                {
+                    ToPhysicalAddress = request.tophysicaladdress,
+                    MessageRedirectId = messageRedirectId
+                };
+
+                Bus.SendLocal(message);
+
+                return Negotiate.WithStatusCode(HttpStatusCode.Accepted);
+            };
+
+            Delete["/redirects/{messageredirectid}"] = parameters =>
+            {
+
+                Guid messageRedirectId = parameters.messageredirectid;
+
+                Bus.SendLocal(new RemoveMessageRedirect
+                {
+                    MessageRedirectId = messageRedirectId
+                });
+
+                return HttpStatusCode.Accepted;
+            };
+
+            Head["/redirects"] = _ =>
+            {
+                using (var session = Store.OpenSession())
+                {
+                    RavenQueryStatistics stats;
+
+                    session
+                        .Query<MessageRedirect>()
+                        .Statistics(out stats);
+
+                    return Negotiate
+                        .WithTotalCount(stats)
+                        .WithEtagAndLastModified(stats);
+                }
+            };
+
+            Get["/redirects"] = _ =>
+            {
+                using (var session = Store.OpenSession())
+                {
+                    RavenQueryStatistics stats;
+
+                    var queryResult = session.Query<MessageRedirect>()
+                        .Statistics(out stats)
+                        .Sort(Request)
+                        .Paging(Request)
+                        .ToArray()
+                        .Select(r => new
+                        {
+                            MessageRedirectId = MessageRedirect.GetMessageRedirectIdFromDocumentId(r.Id),
+                            r.FromPhysicalAddress,
+                            r.ToPhysicalAddress,
+                            LastModified = session.Advanced.GetMetadataFor(r)["Last-Modified"].Value<DateTime>()
+                        });
+
+                    return Negotiate
+                        .WithModel(queryResult)
+                        .WithPagingLinksAndTotalCount(stats, Request)
+                        .WithEtagAndLastModified(stats);
+                }
+            };
+        }
+
+        private Negotiator ValidateRedirectRequest(MessageRedirectRequest request, bool checkForExisting = true)
```

Comment:
I wonder if there is much point doing this validation here.
The data could have changed once we actually doing the update in Raven so maybe just accept the data and report an error later on via a `EventLogMappingDefinition<MessageRedirectCreated>` and `EventLogMappingDefinition<MessageRedirectChanged>`